### PR TITLE
Refactor Database Seeding Process

### PR DESCRIPTION
**Pull request by Martinez-Emmanuel**
Good day, FLSS team. This PR is a refactor of the seeding process in our database using LARAVEL. To understand why I did this despite not a major requirement in our process:

- In my last pull request #85 , I highlighted the use of indexing across our database to optimize the query results
- Right now, we have two data seeds: (1) the sample data defined in our `seeders` directory that is used when you run `php artisan db:seed` in the local development and (2) the actual, initial data that you encoded in our deployed database used in online beta testing. So basically we have two different datasets in local and deployment

I mentioned in our past meetings that for some reason, I sometimes encounter bugs when I switch between the seeder data in our LARAVEL project and the actual encoded data (which I manually export from our hostinger db and import it to our local). This could be due to ID or table structure discrepancies, etc.

And when I added indexing to our database, that would mean that all the actual, initial data that you encoded needs to be re-written from scratch so that it would benefit from the new indexed columns. Although we can run sql queries to modify the table directly, based on my testing, this causes unstable behavior.

I know encoding the curriculum from scratch would be time-consuming for our data encoders so I tried to streamline this process. Below I will outline how I did it so you can add/remove/modify/debug it in the future:

### Step 1: In phpMyAdmin, export a table into a `.csv` format:

image
Unfortunately, phpMyAdmin does not give an option  to export entire database to a .csv - this needs to be done per table. However since I already exported all our tables, you don't need to do this in the future.

### Step 2: Add the .csv file inside the `database/seeders/csv` directory. This directory contains all our csv files.
image

### Step 3: Inside the `csv/` directory, I created a script named `csv_to_array.php`. This script is fairly simple but powerful: It will read the csv file, treat the first row as header, and the rest are treated as records. Then each row will be converted into an array.

```php
    $csv = Reader::createFromPath($csvPath, 'r');
    $csv->setHeaderOffset(0);
    $records = $csv->getRecords();

    $data = [];
    foreach ($records as $record) {
        $data[] = $record;
    }
    return $data;
``` 
For this I use an awesome library named `league/csv`. It's a lightweight but well-maintained library for reading and manipulating CSV files. You can check it out here: [https://csv.thephpleague.com/](https://csv.thephpleague.com/)

### Step 4: Use the `csv_to_array.php` script in the seeder files. All you need to do is:

4.1 Import the converter script into the seeder:
```php 
use Database\Seeders\Csv\CsvToArray;
```

4.2 Inside the `run()` function, add this:
```php 
$csvPath = database_path('seeders/csv/curricula.csv');
```

4.3 Initialize the csv to array conversion:
```php 
$csvData = CsvToArray::convert($csvPath);
```

4.4 Loop over each line
```php 
$dataToInsert = [];

foreach ($csvData as $record) {
    $dataToInsert[] = [
        'id' => $record['id'],
        'user_id' => $record['user_id'],
        'faculty_type' => $record['faculty_type'],
        'faculty_units' => $record['faculty_units'],
        'created_at' => $record['created_at'] ?? null,
        'updated_at' => $record['updated_at'] ?? null,
    ];
}
```
4.5 And then insert it into our database

```php
  DB::table('schedules')->insert($dataToInsert);
  $this->command->info('Schedules table seeded successfully!');
```

If you run the seeder on terminal, this should be the output:

image

I already did this to all of our 21 seeders so all you need to do is run `php artisan migrate:fresh --seed` and now your local development has now the latest datasets that you encoded in the deployment.

**TL; DR: I refactored our seeding process to enhance the developer experience, apply the advantages of indexing to the actual, existing database without manually writing them again from the scratch, and so that we are all working in the actual, real-world data given to us during development this day and onwards. Also, this is a better and more correct version of how seeding is done in the real-world, so it's nice to be implemented in our system as well.**

Again, test this on your local machines and let me know if there are any issues. Kindly double check if the data in our csv, particularly in the curriculum is correct @Rasquero-ViaClariz @Naoe-Adrian. I imported them from our .sql file so if there are any discrepancies, it's very likely an error in the encoding. You can simply edit our .csv files if you find any incorrect data (like course name etc.) or change it in the UI, then import the updated .csv file again - it's up to you. If you are simply updating the data, all you need to do is Step 1 and 2. The script and seeder will work as is.

### CODE REVIEW

**Review by Rasquero-ViaClariz — COMMENTED** [review]
Thank you for refactoring the seeding process and providing clear steps for future updates. The use of CSV files and the script to streamline data handling is efficient and well-explained. I’ll test this locally and double-check the curriculum data for any discrepancies. I’ll let you know if I find issues or errors during the process. Great work on improving the workflow!

**Comment by Naoe-Adrian** [comment]
LGTM THANKS ILL STUDY THIS LATER
